src/mpl2/src/SACoreSoftMacro.cpp

```diff
@@ -0,0 +1,877 @@
+  // again. So when we need to generate the final floorplan out,
+  // we need to call PackFloorplan again at the end of SA process
+  if (action_id_ == 5)
+    macros_[macro_id_] = pre_macros_[macro_id_];
+  else if (action_id_ == 1)
+    pos_seq_ = pre_pos_seq_;
+  else if (action_id_ == 2)
+    neg_seq_ = pre_neg_seq_;
+  else {
+    pos_seq_ = pre_pos_seq_;
+    neg_seq_ = pre_neg_seq_;
+  }
+
+  width_ = pre_width_;
+  height_ = pre_height_;
+  outline_penalty_ = pre_outline_penalty_;
+  wirelength_ = pre_wirelength_;
+  guidance_penalty_ = pre_guidance_penalty_;
+  fence_penalty_ = pre_fence_penalty_;
+  boundary_penalty_ = pre_boundary_penalty_;
+  macro_blockage_penalty_ = pre_macro_blockage_penalty_;
+  notch_penalty_ = pre_notch_penalty_;
+}
+
+void SACoreSoftMacro::Initialize()
+{
+  std::vector<float> outline_penalty_list;
+  std::vector<float> wirelength_list;
+  std::vector<float> guidance_penalty_list;
+  std::vector<float> fence_penalty_list;
+  std::vector<float> boundary_penalty_list;
+  std::vector<float> macro_blockage_penalty_list;
+  std::vector<float> notch_penalty_list;
+  std::vector<float> area_penalty_list;
+  std::vector<float> width_list;
+  std::vector<float> height_list;
+  for (int i = 0; i < num_perturb_per_step_; i++) {
+    Perturb();
+    // store current penalties
+    width_list.push_back(width_);
+    height_list.push_back(height_);
+    area_penalty_list.push_back(width_ * height_);
+    outline_penalty_list.push_back(outline_penalty_);
+    wirelength_list.push_back(wirelength_);
+    guidance_penalty_list.push_back(guidance_penalty_);
+    fence_penalty_list.push_back(fence_penalty_);
+    boundary_penalty_list.push_back(boundary_penalty_);
+    macro_blockage_penalty_list.push_back(macro_blockage_penalty_);
+    notch_penalty_list.push_back(notch_penalty_);
+  }
+
+  norm_area_penalty_ = CalAverage(area_penalty_list);
+  norm_outline_penalty_ = CalAverage(outline_penalty_list);
+  norm_wirelength_ = CalAverage(wirelength_list);
+  norm_guidance_penalty_ = CalAverage(guidance_penalty_list);
+  norm_fence_penalty_ = CalAverage(fence_penalty_list);
+  norm_boundary_penalty_ = CalAverage(boundary_penalty_list);
+  norm_macro_blockage_penalty_ = CalAverage(macro_blockage_penalty_list);
+  // norm_notch_penalty_    = CalAverage(notch_penalty_list);
+  norm_notch_penalty_ = outline_width_ * outline_height_;
+
+  // Calculate initial temperature
+  std::vector<float> cost_list;
+  for (int i = 0; i < outline_penalty_list.size(); i++) {
+    width_ = width_list[i];
+    height_ = height_list[i];
+    outline_penalty_ = outline_penalty_list[i];
+    wirelength_ = wirelength_list[i];
+    guidance_penalty_ = guidance_penalty_list[i];
+    fence_penalty_ = fence_penalty_list[i];
+    boundary_penalty_ = boundary_penalty_list[i];
+    macro_blockage_penalty_ = macro_blockage_penalty_list[i];
+    notch_penalty_ = notch_penalty_list[i];
+    cost_list.push_back(CalNormCost());
```

Comment:
Why aren't you passing these as arguments to CalNormCost rather than through member vars?